tracer/src/Datadog.Trace/Sampling/TraceSampler.cs

```diff
@@ -137,5 +87,68 @@ private SamplingDecision MakeSamplingDecision(Span span, float rate, string mech
 
             return new SamplingDecision(priority, mechanism, rate, limiterRate);
         }
+
+        public class Builder(IRateLimiter limiter)
+        {
+            private readonly IRateLimiter _limiter = limiter;
+            private readonly List<ISamplingRule> _rules = [];
+            private AgentSamplingRule? _agentSamplingRule;
+
+            public TraceSampler Build()
+            {
+                return new TraceSampler(_limiter, _rules, _agentSamplingRule);
+            }
+
+            /// <summary>
+            /// Register a new sampling rule. To register a <see cref="AgentSamplingRule"/>,
+            /// use <see cref="RegisterAgentSamplingRule"/> instead.
+            /// </summary>
+            /// <remarks>
+            /// The order that rules are registered is important, as they are evaluated in order.
+            /// The first rule that matches will be used to determine the sampling rate.
+            /// </remarks>
+            public void RegisterRule(ISamplingRule rule)
+            {
+                _rules.Add(rule);
+            }
+
+            /// <summary>
+            /// Register new sampling rules. To register a <see cref="AgentSamplingRule"/>,
+            /// use <see cref="RegisterAgentSamplingRule"/> instead.
+            /// </summary>
+            /// <remarks>
+            /// The order that rules are registered is important, as they are evaluated in order.
+            /// The first rule that matches will be used to determine the sampling rate.
+            /// </remarks>
+            public void RegisterRules(IEnumerable<ISamplingRule> rules)
+            {
+                _rules.AddRange(rules);
+            }
+
+            /// <summary>
+            /// Register a new agent sampling rule. This rule should be registered last,
+            /// after any calls to <see cref="RegisterRule"/> or <see cref="RegisterRules"/>.
+            /// </summary>
+            /// <remarks>
+            /// The order that rules are registered is important, as they are evaluated in order.
+            /// The first rule that matches will be used to determine the sampling rate.
+            /// </remarks>
+            public void RegisterAgentSamplingRule(AgentSamplingRule rule)
+            {
+                // only register the one AgentSamplingRule
+                if (Interlocked.Exchange(ref _agentSamplingRule, rule) == null)
+                {
+                    // keep a reference to this rule so we can call SetDefaultSampleRates() later
+                    // to update the agent sampling rates
+                    _agentSamplingRule = rule;
```

Comment:
```suggestion
```

Just seeing this but I don't think we need this set as we are doing a `Interlocked.Exchange` on it?

Was already there though.

Reply:
Oh yeah, I didn't read this code at all 😅 I agree, I don't see why the set is necessary 👍